### PR TITLE
feat(modules/nixvim): add color scheme

### PR DESCRIPTION
**Pull request by trueNAHO**
# About

Closes: https://github.com/danth/stylix/issues/153

# Need Help

Currently, this PR does not integrate with [NixVim](https://github.com/nix-community/nixvim)'s generated configuration file. This might be related to NixVim not being a Home Manager module.

### CODE REVIEW

**Review by danth — CHANGES_REQUESTED** [review]
`programs.nixvim` is not available by default so this will cause an error for configs without NixVim imported.

Also, the module file needs to be called `hm.nix`, `nixos.nix` or `darwin.nix` to be loaded properly. It looks like NixVim can be installed on any of those, so we should support them all. If the modules are identical then each file can just `import ./nixvim.nix`, or symlinks might work instead.

**Comment by trueNAHO** [comment]
> `programs.nixvim` is not available by default so this will cause an error for configs without NixVim imported.

Should I add NixVim to the `flake.nix` inputs?

**Comment by danth** [comment]
That would cause it to be installed for a lot of people who don't want to use it.

**Comment by trueNAHO** [comment]
> That would cause it to be installed for a lot of people who don't want to use it.

Do you have any other proposals?

Conditionally calling `fetchFromGithub` would not be ideal as we lose major benefits that Flakes provide. Hence, we either add NixVim to the Flake inputs, or we do not add it as an input at all. In the latter case, we should assert NixVim being available in the scope that uses Stylix as input. 

Do you have any ideas on how to check if NixVim is available?

**Comment by danth** [comment]
This was something we discussed back when the Hyprland module was not part of Home Manager. In short we came to the decision not to add anything to Stylix which relies on an external module being imported, which NixVim does.

If you want to experiment anyway, `options.programs?nixvim` would return `true` when `programs.nixvim` exists as an option. (`options` can be loaded at the top of the file similar to `pkgs` and `config`.)

File: modules/nixvim/nixvim.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+  config,
+  lib,
+  ...
+}: {
+  options.stylix.targets.nixvim.enable =
+    config.lib.stylix.mkEnableTarget "nixvim" true;
+
+  config = lib.mkIf config.stylix.targets.nixvim.enable {
```
Comment (trueNAHO) [inline modules/nixvim/nixvim.nix:9]: What about the following assertion:

```suggestion
  config = lib.mkIf config.stylix.targets.nixvim.enable {
    assertions = [
      {
        assertion = options.programs?nixvim;

        message = ''
          'stylix.targets.nixvim.enable' is set to
          ${config.stylix.targets.nixvim.enable}, but the external
          'options.programs.nixvim' dependency is unavailable:
          https://github.com/nix-community/nixvim
        '';
      }
    ];
```

Note that I have not actually tested this.

Comment (danth) [reply modules/nixvim/nixvim.nix:9]: With `mkIf`, evaluation still fails when the option doesn't exist, even if `config.stylix.targets.nixvim.enable` is set to `false`

Comment (trueNAHO) [reply modules/nixvim/nixvim.nix:9]: The following should gracefully skip evaluation if `config.stylix.targets.nixvim` is not available:

```suggestion
  config = let
    targets = config.stylix.targets;
  in lib.mkIf (targets?nixvim.enable && targets.nixivm.enable) {
```

Note that I have not actually tested this.

**Comment by trueNAHO** [comment]
> Also, the module file needs to be called `hm.nix`, `nixos.nix` or `darwin.nix` to be loaded properly. It looks like NixVim can be installed on any of those, so we should support them all. If the modules are identical then each file can just `import ./nixvim.nix`, or symlinks might work instead.

I will work on this once I have more free time.

**Comment by willemml** [comment]
I have fixed the issues with this PR in this fork: https://github.com/willemml/stylix/tree/feat/modules/nixvim
See commit https://github.com/willemml/stylix/commit/8bb621eb27ef67087f84d6e9b93a7649152b06b2.

I have tested that the build succeeds without installing nixvim. Colors are applied correctly. The only thing I cannot figure out is transparency, the base nixvim theme follows my terminal's transparency settings while this stylix module sets the background to a solid colour.

Assuming these fixes/changes are acceptable should I submit this as a separate PR to the main danth/stylix repo?

**Comment by trueNAHO** [comment]
> I have tested that the build succeeds without installing nixvim. Colors are applied correctly.

I have not tested your PR, but the code looks good.

> I have added an option for terminal transparency passthrough in a separate branch, I am unsure if this follows stylix module guidelines but here it is: https://github.com/willemml/stylix/tree/feat/modules/nixvim-transparency ([willemml@12108f4](https://github.com/willemml/stylix/commit/12108f41b1fd4790d7743f0b2478bb534114b2c9)).

I don't see how more options would be problematic. @danth, what do you think?

> Assuming these fixes/changes are acceptable should I submit this as a separate PR to the main danth/stylix repo?

Regardless of transparency being accepted, your PR already supersedes mine. Feel free to open a new PR, and link to this one so that I can close it.

**Pull request closed without merging**